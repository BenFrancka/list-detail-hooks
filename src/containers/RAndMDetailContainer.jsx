import React, { useState, useEffect } from 'react';
import { Link, useParams } from 'react-router-dom';
// eslint-disable-next-line max-len
import CharacterDetailDisplay from '../components/characters/CharacterDetailDisplay';
import { fetchCharacterById } from '../services/RAndMApi';

const RAndMDetailContainer = () => {
  const { id } = useParams();
  const [loading, setLoading] = useState(true);
  const [character, setCharacter] = useState({});

  useEffect(() => {
    fetchCharacterById(id)
      .then((character) => setCharacter(character))
      .finally(() => setLoading(false));
  }, [id]);

  return (
    <main>
      <Link to="/">Portal to Homepage</Link>
      {loading ? (
        <img
          src="https://thumbs.gfycat.com/SoftEarnestGyrfalcon-max-1mb.gif"
          alt="loading spinner"
        />
      ) : (
        <CharacterDetailDisplay
          image={character.image}
          name={character.name}
          status={character.status}
          species={character.species}
        />
      )}
    </main>
  );
};

export default RAndMDetailContainer;
