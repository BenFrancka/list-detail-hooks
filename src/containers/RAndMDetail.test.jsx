import React from 'react';
import { render, screen } from '@testing-library/react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';
import { MemoryRouter } from 'react-router-dom';
import RAndMDetailContainer from './RAndMDetailContainer';
import DetailData from '../fixtures/DetailData.json';

const server = setupServer(
  rest.get('https://rickandmortyapi.com/api/character/:id', (req, res, ctx) => {
    return res(ctx.json(DetailData));
  })
);

describe('RAndMDetailContainer', () => {
  beforeAll(() => server.listen());
  afterAll(() => server.close());

  // eslint-disable-next-line max-len
  it('displays a detail view of a single Rick and Morty Character', async () => {
    const { container } = render(
      <MemoryRouter initialEntries={['/2']}>
        <RAndMDetailContainer />
      </MemoryRouter>
    );

    screen.getAllByAltText('loading spinner');

    await screen.getAllByAltText('Morty Smith');

    const ul = await screen.findByRole('list', { name: 'character-details' });
    expect(ul).not.toBeEmptyDOMElement();
    expect(container).toMatchSnapshot();
  });
});
