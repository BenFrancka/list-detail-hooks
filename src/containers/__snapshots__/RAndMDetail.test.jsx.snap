// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RAndMDetailContainer displays a detail view of a single Rick and Morty Character 1`] = `
<div>
  <main>
    <a
      href="/"
    >
      Portal to Homepage
    </a>
    <ul
      aria-label="character details"
    >
      <li>
        <img
          alt="Morty Smith"
          src="https://rickandmortyapi.com/api/character/avatar/2.jpeg"
        />
      </li>
      <li>
        Morty Smith
      </li>
      <li>
        Alive
      </li>
      <li>
        Human
      </li>
    </ul>
  </main>
</div>
`;
